# EDA-Cleaning: Part II

In [1]:
import os
import pandas as pd
import numpy as np

### Check the contents of our current working directory

In [2]:
os.listdir()

['.git',
 '.ipynb_checkpoints',
 'books_rating_project_EDA_3.ipynb',
 'books_rating_project_EDA_Cleaning_1.ipynb',
 'books_rating_project_EDA_Cleaning_2.ipynb',
 'data',
 'README.md']

### Check the contents of our data directory

In [3]:
os.listdir('data')

['cleaned_books.csv',
 'uncleaned_books.csv',
 'uncleaned_books.csv_bad_lines.txt',
 'working_books.csv']

### Start of Cleaning II

In [4]:
folder = 'data/'
books_csv = 'working_books.csv'
f_path = folder + books_csv

In [5]:
df_ratings = pd.read_csv(f_path)

In [6]:
df_ratings.head()

bookID                                              title  \
0       1  Harry Potter and the Half-Blood Prince (Harry ...   
1       2  Harry Potter and the Order of the Phoenix (Har...   
2       4  Harry Potter and the Chamber of Secrets (Harry...   
3       5  Harry Potter and the Prisoner of Azkaban (Harr...   
4       8  Harry Potter Boxed Set  Books 1-5 (Harry Potte...   

                      authors  average_rating        isbn         isbn13  \
0  J.K. Rowling/Mary GrandPré            4.57  0439785960  9780439785969   
1  J.K. Rowling/Mary GrandPré            4.49  0439358078  9780439358071   
2                J.K. Rowling            4.42  0439554896  9780439554893   
3  J.K. Rowling/Mary GrandPré            4.56  043965548X  9780439655484   
4  J.K. Rowling/Mary GrandPré            4.78  0439682584  9780439682589   

  language_code    num_pages  ratings_count  text_reviews_count  \
0           eng          652        2095690               27591   
1           eng          870        2153167               29221   
2           eng          352           6333                 244   
3           eng          435        2339585               36325   
4           eng         2690          41428                 164   

  publication_date        publisher  
0        9/16/2006  Scholastic Inc.  
1         9/1/2004  Scholastic Inc.  
2        11/1/2003       Scholastic  
3         5/1/2004  Scholastic Inc.  
4        9/13/2004       Scholastic

In [7]:
df_ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11127 entries, 0 to 11126
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   bookID              11127 non-null  int64  
 1   title               11127 non-null  object 
 2   authors             11127 non-null  object 
 3   average_rating      11127 non-null  float64
 4   isbn                11127 non-null  object 
 5   isbn13              11127 non-null  int64  
 6   language_code       11127 non-null  object 
 7     num_pages         11127 non-null  int64  
 8   ratings_count       11127 non-null  int64  
 9   text_reviews_count  11127 non-null  int64  
 10  publication_date    11127 non-null  object 
 11  publisher           11127 non-null  object 
dtypes: float64(1), int64(5), object(6)
memory usage: 1.0+ MB


### Move the target column ('average_rating') to the end of the table

In [8]:
lst_headers = list(df_ratings.columns)

In [9]:
target_index = lst_headers.index('average_rating')

lst_headers.append(lst_headers.pop(target_index))
lst_headers


['bookID',
 'title',
 'authors',
 'isbn',
 'isbn13',
 'language_code',
 '  num_pages',
 'ratings_count',
 'text_reviews_count',
 'publication_date',
 'publisher',
 'average_rating']

In [10]:
df_ratings = df_ratings[lst_headers]
df_ratings.head()

bookID                                              title  \
0       1  Harry Potter and the Half-Blood Prince (Harry ...   
1       2  Harry Potter and the Order of the Phoenix (Har...   
2       4  Harry Potter and the Chamber of Secrets (Harry...   
3       5  Harry Potter and the Prisoner of Azkaban (Harr...   
4       8  Harry Potter Boxed Set  Books 1-5 (Harry Potte...   

                      authors        isbn         isbn13 language_code  \
0  J.K. Rowling/Mary GrandPré  0439785960  9780439785969           eng   
1  J.K. Rowling/Mary GrandPré  0439358078  9780439358071           eng   
2                J.K. Rowling  0439554896  9780439554893           eng   
3  J.K. Rowling/Mary GrandPré  043965548X  9780439655484           eng   
4  J.K. Rowling/Mary GrandPré  0439682584  9780439682589           eng   

     num_pages  ratings_count  text_reviews_count publication_date  \
0          652        2095690               27591        9/16/2006   
1          870        2153167               29221         9/1/2004   
2          352           6333                 244        11/1/2003   
3          435        2339585               36325         5/1/2004   
4         2690          41428                 164        9/13/2004   

         publisher  average_rating  
0  Scholastic Inc.            4.57  
1  Scholastic Inc.            4.49  
2       Scholastic            4.42  
3  Scholastic Inc.            4.56  
4       Scholastic            4.78

In [11]:
df_ratings.describe()

bookID        isbn13     num_pages  ratings_count  \
count  11127.000000  1.112700e+04  11127.000000   1.112700e+04   
mean   21310.938887  9.759888e+12    336.376921   1.793641e+04   
std    13093.358023  4.428964e+11    241.127305   1.124794e+05   
min        1.000000  8.987060e+09      0.000000   0.000000e+00   
25%    10287.000000  9.780345e+12    192.000000   1.040000e+02   
50%    20287.000000  9.780586e+12    299.000000   7.450000e+02   
75%    32104.500000  9.780873e+12    416.000000   4.993500e+03   
max    45641.000000  9.790008e+12   6576.000000   4.597666e+06   

       text_reviews_count  average_rating  
count        11127.000000    11127.000000  
mean           541.854498        3.933631  
std           2576.176608        0.352445  
min              0.000000        0.000000  
25%              9.000000        3.770000  
50%             46.000000        3.960000  
75%            237.500000        4.135000  
max          94265.000000        5.000000

In [12]:
df_ratings.describe(include=object)

title       authors        isbn language_code publication_date  \
count       11127         11127       11127         11127            11127   
unique      10352          6643       11127            27             3679   
top     The Iliad  Stephen King  0439785960           eng        10/1/2005   
freq            9            40           1          8911               56   

       publisher  
count      11127  
unique      2292  
top      Vintage  
freq         318

### bookID and isbn13 are both listed as int64, but they seem to be better off as categorical data, so we adjust them to type str

In [13]:
df_ratings = df_ratings.astype({'bookID': str, 'isbn13': str})
df_ratings.describe(include=object)

bookID      title       authors        isbn         isbn13  \
count   11127      11127         11127       11127          11127   
unique  11127      10352          6643       11127          11127   
top         1  The Iliad  Stephen King  0439785960  9780439785969   
freq        1          9            40           1              1   

       language_code publication_date publisher  
count          11127            11127     11127  
unique            27             3679      2292  
top              eng        10/1/2005   Vintage  
freq            8911               56       318

In [14]:
df_ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11127 entries, 0 to 11126
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   bookID              11127 non-null  object 
 1   title               11127 non-null  object 
 2   authors             11127 non-null  object 
 3   isbn                11127 non-null  object 
 4   isbn13              11127 non-null  object 
 5   language_code       11127 non-null  object 
 6     num_pages         11127 non-null  int64  
 7   ratings_count       11127 non-null  int64  
 8   text_reviews_count  11127 non-null  int64  
 9   publication_date    11127 non-null  object 
 10  publisher           11127 non-null  object 
 11  average_rating      11127 non-null  float64
dtypes: float64(1), int64(3), object(8)
memory usage: 1.0+ MB


### publication_date is currently a string, but we may want to try and do datetime operations, so we'll change the publication_date to datetime

In [15]:
df_ratings['publication_date'] = pd.to_datetime(df_ratings['publication_date'])

ValueError: day is out of range for month present at position 8177

### At least one of the dates was not able to be converted due to value error. We do a quick spot check on the dates to see if a date was entered as something other than dd-mm-yyyy format

In [16]:
df_ratings[(df_ratings.language_code != 'eng') & (df_ratings.language_code != 'en-US') & (df_ratings.language_code != 'en-CA')
          & (df_ratings.language_code != 'en-GB')]

bookID                                              title  \
50        80                     La Place de la Concorde Suisse   
109      201        Una arruga en el tiempo – A Wrinkle in Time   
146      324                               Cien años de soledad   
150      337                 El perfume: Historia de un asesino   
201      466                 Timbuktu / Leviathan / Moon Palace   
...      ...                                                ...   
11113  45617  O Cavalo e o Seu Rapaz (As Crónicas de Nárnia ...   
11114  45623   O Sobrinho do Mágico (As Crónicas de Nárnia  #1)   
11115  45625  A Viagem do Caminheiro da Alvorada (As Crónica...   
11116  45626     O Príncipe Caspian (As Crónicas de Nárnia  #4)   
11122  45641                        Las aventuras de Tom Sawyer   

                                           authors        isbn         isbn13  \
50                                     John McPhee  0374519323  9780374519322   
109                              Madeleine L'Engle  0606105263  9780606105262   
146                         Gabriel García Márquez  0785950109  9780785950103   
150                                Patrick Süskind  8432216062  9788432216060   
201                                    Paul Auster  2742741461  9782742741465   
...                                            ...         ...            ...   
11113  C.S. Lewis/Pauline Baynes/Ana Falcão Bastos  9722330551  9789722330558   
11114  C.S. Lewis/Pauline Baynes/Ana Falcão Bastos  9722329987  9789722329989   
11115  C.S. Lewis/Pauline Baynes/Ana Falcão Bastos  9722331329  9789722331326   
11116  C.S. Lewis/Pauline Baynes/Ana Falcão Bastos  9722330977  9789722330978   
11122                                   Mark Twain  8497646983  9788497646987   

      language_code    num_pages  ratings_count  text_reviews_count  \
50              fre          160            698                  52   
109             spa          205              6                   1   
146             spa          448             63                   7   
150             spa          239           4136                 255   
201             fre         1075             21                   1   
...             ...          ...            ...                 ...   
11113           por          160            207                  16   
11114           por          147            396                  37   
11115           por          176            161                  14   
11116           por          160            215                  11   
11122           spa          272            113                  12   

      publication_date                  publisher  average_rating  
50            4/1/1994  Farrar  Straus and Giroux            3.92  
109           6/1/1984           Turtleback Books            4.00  
146           1/1/1990          French & European            4.07  
150           6/1/2002                     Booket            4.02  
201          11/7/2002                  Actes Sud            4.38  
...                ...                        ...             ...  
11113        8/15/2003         Editorial Presença            3.92  
11114         4/8/2003         Editorial Presença            4.04  
11115         9/1/2004         Editorial Presença            4.09  
11116       10/11/2003         Editorial Presença            3.97  
11122        5/28/2006              Edimat Libros            3.91  

[586 rows x 12 columns]

### Spot check of data does not immediately show the problem. We'll change the strings to dates in a new column, turn the errors to 'null' values, and use that as a way to fix our problem data

In [17]:
df_ratings['publication_date_fixed'] = pd.to_datetime(df_ratings['publication_date'], errors='coerce')

In [18]:
df_ratings.head(1)

bookID                                              title  \
0      1  Harry Potter and the Half-Blood Prince (Harry ...   

                      authors        isbn         isbn13 language_code  \
0  J.K. Rowling/Mary GrandPré  0439785960  9780439785969           eng   

     num_pages  ratings_count  text_reviews_count publication_date  \
0          652        2095690               27591        9/16/2006   

         publisher  average_rating publication_date_fixed  
0  Scholastic Inc.            4.57             2006-09-16

In [19]:
df_ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11127 entries, 0 to 11126
Data columns (total 13 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   bookID                  11127 non-null  object        
 1   title                   11127 non-null  object        
 2   authors                 11127 non-null  object        
 3   isbn                    11127 non-null  object        
 4   isbn13                  11127 non-null  object        
 5   language_code           11127 non-null  object        
 6     num_pages             11127 non-null  int64         
 7   ratings_count           11127 non-null  int64         
 8   text_reviews_count      11127 non-null  int64         
 9   publication_date        11127 non-null  object        
 10  publisher               11127 non-null  object        
 11  average_rating          11127 non-null  float64       
 12  publication_date_fixed  11125 non-null  dateti

### There appear to be two strings that were somehow formatted incorrectly

In [20]:
df_ratings[(df_ratings['publication_date_fixed'].isnull())]

bookID                                              title  \
8177   31373  In Pursuit of the Proper Sinner (Inspector Lyn...   
11094  45531         Montaillou  village occitan de 1294 à 1324   

                                               authors        isbn  \
8177                                 Elizabeth  George  0553575104   
11094  Emmanuel Le Roy Ladurie/Emmanuel Le Roy-Ladurie  2070323285   

              isbn13 language_code    num_pages  ratings_count  \
8177   9780553575101           eng          718          10608   
11094  9782070323289           fre          640             15   

       text_reviews_count publication_date       publisher  average_rating  \
8177                  295       11/31/2000    Bantam Books            4.10   
11094                   2        6/31/1982  Folio histoire            3.96   

      publication_date_fixed  
8177                     NaT  
11094                    NaT

### Both dates are incorrect, listing the publication date as the 31st of months that only have 30 days each. A quick check of the book entries on the goodreads website (add weblinks for the two books here) shows that book 31373 was published on 10/31/2000 and book 45531 was published on 6/30/1982. We'll fix these two entries in the code below

In [21]:
df_ratings.at[8177, 'publication_date']

'11/31/2000'

In [22]:
df_ratings.at[8177, 'publication_date'] = '10/31/2000'
df_ratings.at[8177, 'publication_date']

'10/31/2000'

In [23]:
df_ratings.at[11094, 'publication_date']

'6/31/1982'

In [24]:
df_ratings.at[11094, 'publication_date'] = '6/30/1982'
df_ratings.at[11094, 'publication_date']

'6/30/1982'

### And now we should be able to convert our original column to type date

In [25]:
df_ratings.publication_date = pd.to_datetime(df_ratings.publication_date)

In [26]:
df_ratings.head()

bookID                                              title  \
0      1  Harry Potter and the Half-Blood Prince (Harry ...   
1      2  Harry Potter and the Order of the Phoenix (Har...   
2      4  Harry Potter and the Chamber of Secrets (Harry...   
3      5  Harry Potter and the Prisoner of Azkaban (Harr...   
4      8  Harry Potter Boxed Set  Books 1-5 (Harry Potte...   

                      authors        isbn         isbn13 language_code  \
0  J.K. Rowling/Mary GrandPré  0439785960  9780439785969           eng   
1  J.K. Rowling/Mary GrandPré  0439358078  9780439358071           eng   
2                J.K. Rowling  0439554896  9780439554893           eng   
3  J.K. Rowling/Mary GrandPré  043965548X  9780439655484           eng   
4  J.K. Rowling/Mary GrandPré  0439682584  9780439682589           eng   

     num_pages  ratings_count  text_reviews_count publication_date  \
0          652        2095690               27591       2006-09-16   
1          870        2153167               29221       2004-09-01   
2          352           6333                 244       2003-11-01   
3          435        2339585               36325       2004-05-01   
4         2690          41428                 164       2004-09-13   

         publisher  average_rating publication_date_fixed  
0  Scholastic Inc.            4.57             2006-09-16  
1  Scholastic Inc.            4.49             2004-09-01  
2       Scholastic            4.42             2003-11-01  
3  Scholastic Inc.            4.56             2004-05-01  
4       Scholastic            4.78             2004-09-13

### And now we drop our helper column

In [27]:
cols = list(df_ratings.keys())

In [28]:
cols.remove(cols[len(cols)-1])
cols

['bookID',
 'title',
 'authors',
 'isbn',
 'isbn13',
 'language_code',
 '  num_pages',
 'ratings_count',
 'text_reviews_count',
 'publication_date',
 'publisher',
 'average_rating']

In [29]:
df_ratings = df_ratings[cols]

In [30]:
df_ratings.head(1)

bookID                                              title  \
0      1  Harry Potter and the Half-Blood Prince (Harry ...   

                      authors        isbn         isbn13 language_code  \
0  J.K. Rowling/Mary GrandPré  0439785960  9780439785969           eng   

     num_pages  ratings_count  text_reviews_count publication_date  \
0          652        2095690               27591       2006-09-16   

         publisher  average_rating  
0  Scholastic Inc.            4.57

In [31]:
df_ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11127 entries, 0 to 11126
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   bookID              11127 non-null  object        
 1   title               11127 non-null  object        
 2   authors             11127 non-null  object        
 3   isbn                11127 non-null  object        
 4   isbn13              11127 non-null  object        
 5   language_code       11127 non-null  object        
 6     num_pages         11127 non-null  int64         
 7   ratings_count       11127 non-null  int64         
 8   text_reviews_count  11127 non-null  int64         
 9   publication_date    11127 non-null  datetime64[ns]
 10  publisher           11127 non-null  object        
 11  average_rating      11127 non-null  float64       
dtypes: datetime64[ns](1), float64(1), int64(3), object(7)
memory usage: 1.0+ MB


In [32]:
cleaned_books_path = folder + 'cleaned_books.csv'

In [33]:
df_ratings.to_csv(cleaned_books_path, encoding='utf-8', index=False)

In [34]:
os.listdir('data')

['cleaned_books.csv',
 'uncleaned_books.csv',
 'uncleaned_books.csv_bad_lines.txt',
 'working_books.csv']

In [35]:
df_test = pd.read_csv('data/cleaned_books.csv')
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11127 entries, 0 to 11126
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   bookID              11127 non-null  int64  
 1   title               11127 non-null  object 
 2   authors             11127 non-null  object 
 3   isbn                11127 non-null  object 
 4   isbn13              11127 non-null  int64  
 5   language_code       11127 non-null  object 
 6     num_pages         11127 non-null  int64  
 7   ratings_count       11127 non-null  int64  
 8   text_reviews_count  11127 non-null  int64  
 9   publication_date    11127 non-null  object 
 10  publisher           11127 non-null  object 
 11  average_rating      11127 non-null  float64
dtypes: float64(1), int64(5), object(6)
memory usage: 1.0+ MB


In [36]:
?? df_test.to_csv